crypto/rsa_extra/rsassa_pss_asn1.c

```diff
@@ -0,0 +1,317 @@
+      parse_oid(&hash_oid, rsa_pss_hash_functions,
+                OPENSSL_ARRAY_SIZE(rsa_pss_hash_functions), &hash_algor)) {
+    *mga = RSA_MGA_IDENTIFIER_new();
+    if (*mga != NULL) {
+      (*mga)->mask_gen = mgf1;
+      (*mga)->one_way_hash = hash_algor;
+      return 1;
+    }
+  }
+  OPENSSL_PUT_ERROR(RSA, EVP_R_DECODE_ERROR);
+  RSA_ALGOR_IDENTIFIER_free(mgf1);
+  RSA_ALGOR_IDENTIFIER_free(hash_algor);
+  return 0;
+}
+
+// get_context_specific_value sets |*out| to the contents of DER-encoded, ASN.1
+// element (not including tag and length bytes) and advances |seq| over it.
+// When the tag value does not exist, |seq| gets recovered.
+// It returns one when the element exists.
+static int get_context_specific_value(CBS *seq, CBS *out, int index) {
+  unsigned int tag_value = CBS_ASN1_CONTEXT_SPECIFIC | CBS_ASN1_CONSTRUCTED | index;
+  CBS seq_cp = {seq->data, seq->len};
+  if (CBS_get_asn1(seq, out, tag_value)) {
+    return 1;
+  } else {
+    // All fields of RSASSA-PSS-params can be absent, which means default.
+    // Recover |seq|.
+    seq->data = seq_cp.data;
+    seq->len = seq_cp.len;
+    return 0;
+  }
+}
+
+// Decode [0] HashAlgorithm of RSASSA-PSS-params: return one on success and zero
+// on failure. See 3.1. https://tools.ietf.org/html/rfc4055#page-7
+static int decode_pss_hash(CBS *seq, RSA_ALGOR_IDENTIFIER **hash_algor) {
+  CBS cs;
+  if (!get_context_specific_value(seq, &cs, 0)) {
+    // HashAlgorithm field can be absent, which means default.
+    return 1;
+  }
+  return decode_one_way_hash(&cs, hash_algor);
+}
+
+// Decode [1] MaskGenAlgorithm of RSASSA-PSS-params: return one on success and
+// zero on failure. See 3.1. https://tools.ietf.org/html/rfc4055#page-7
+static int decode_pss_mask_gen(CBS *seq, RSA_MGA_IDENTIFIER **mga) {
+  CBS cs;
+  if (!get_context_specific_value(seq, &cs, 1)) {
+    // MaskGenAlgorithm field can be absent, which means default.
+    return 1;
+  }
+  return decode_mask_gen(&cs, mga);
+}
+
+static int parse_rsa_int(CBS *cbs, RSA_INTEGER **rsa_int) {
+  int64_t value = 0;
+  if (CBS_get_asn1_int64(cbs, &value) && CBS_len(cbs) == 0) {
+    *rsa_int = RSA_INTEGER_new();
+    if ((*rsa_int) != NULL) {
+      (*rsa_int)->value = value;
+      return 1;
+    }
+  }
+  OPENSSL_PUT_ERROR(RSA, EVP_R_DECODE_ERROR);
+  return 0;
+}
+
+// Decode [2] saltLength of RSASSA-PSS-params
+// See 3.1. https://tools.ietf.org/html/rfc4055#page-7
+static int decode_pss_salt_len(CBS *seq, RSA_INTEGER **salt_len) {
+  CBS cs;
+  if (!get_context_specific_value(seq, &cs, 2)) {
+    // saltLength field can be absent, which means default.
+    return 1;
+  }
+  return parse_rsa_int(&cs, salt_len);
+}
+
+// Decode [3] trailerField of RSASSA-PSS-params
+// See 3.1. https://tools.ietf.org/html/rfc4055#page-7
+static int decode_pss_trailer_field(CBS *seq, RSA_INTEGER **trailer_field) {
+  CBS cs;
+  if (!get_context_specific_value(seq, &cs, 3)) {
+    // Trailer field can be absent, which means default.
```

Comment:
Can we somehow enforce that the field must be there when performing signature validation, per the RFC?

This might be a task longer up the stack though.

Reply:
Pss params decode and encode are not symmetric. The RFC requires some default values must be omit in encode, but decode has to accept both (absent or non-absent).